### PR TITLE
[#90] Tests for GitHub PR ingestion

### LINKED ISSUE #90: [Epic 8] Tests for GitHub PR ingestion
## Description

Write tests for lib/github-PR-tracker.ts covering the core behaviour and edge cases.

## Acceptance Criteria

- [ ] Only PRs merged within the weekly window are ingested
- [ ] PR author and merged-by identities are resolved and saved correctly
- [ ] PRs with no author or merged-by user are handled without error
- [ ] Duplicate PRs are upserted without error
- [ ] Repos with no merged PRs in the window are handled gracefully

## Dependencies/Blockers

#87

### PR DESCRIPTION
**Pull request by L1ghtn1ng1**
## Summary
- Adds unit tests for `ingestRepoMergedPRs` in `apps/worker/src/lib/github-PR-tracker.ts`
- Covers all 5 acceptance criteria from issue #90

## Test cases
- Returns 0 and skips DB writes when no PRs are merged in window
- Upserts a merged PR and returns correct count
- Resolves author and mergedBy identities and saves them to the PR record
- Handles PRs with no author or mergedBy without throwing
- Passes the correct date range to the GitHub search query

## Test plan
- [ ] Run `pnpm test` in `apps/worker` — all 5 tests pass

Closes #90

### CODE REVIEW

File: apps/worker/src/lib/github-PR-tracker.test.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    await expect(ingestRepoMergedPRs(repo, weekStart, weekEnd)).resolves.toBe(1)
+    expect(db.pRFact.upsert).toHaveBeenCalledWith(
+      expect.objectContaining({
+        create: expect.objectContaining({
+          authorIdentityId: null,
+          mergedByIdentityId: null,
+        }),
+      })
+    )
+  })
```
Comment (kelvinchen8) [inline apps/worker/src/lib/github-PR-tracker.test.ts]: What's wrong: The upsert call has both a create and an update block, and both write authorIdentityId/mergedByIdentityId. The test only asserts on create, so a bug where update drops the identity fields would go completely undetected — which is exactly the scenario this test is meant to guard against.

Comment to leave:

The upsert's update block also writes authorIdentityId and mergedByIdentityId, but only the create block is asserted here. A regression that drops identities from the update path would pass this test silently. Add a second objectContaining check for the update key.

Recommended fix: Add an assertion for the update block alongside the existing one:


expect(db.pRFact.upsert).toHaveBeenCalledWith(
  expect.objectContaining({
    create: expect.objectContaining({
      authorIdentityId: 'author-identity',
      mergedByIdentityId: 'merger-identity',
    }),
    update: expect.objectContaining({
      authorIdentityId: 'author-identity',
      mergedByIdentityId: 'merger-identity',
    }),
  })
)

File: apps/worker/src/lib/github-PR-tracker.test.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    expect(db.pRFact.upsert).toHaveBeenCalledTimes(1)
+    expect(db.pRFact.upsert).toHaveBeenCalledWith(
+      expect.objectContaining({
+        where: { repoId_number: { repoId: 'repo-1', number: 1 } },
+      })
+    )
+  })
+
+  it('resolves author and mergedBy identities and saves them to the PR record', async () => {
+    const { resolveIdentity } = await import('./github-utils.js')
```
Comment (kelvinchen8) [inline apps/worker/src/lib/github-PR-tracker.test.ts]: What's wrong: line 90 and 110 import via import('./github-utils.js') with an explicit .js extension, but the vi.mock declaration at line 11 uses './github-utils' without one. In most Vitest/TypeScript setups these resolve to the same module, but it's fragile — if module resolution changes (e.g. moduleResolution: bundler → node16), the dynamic import could bypass the mock and import the real module, causing the identity assertions to pass spuriously.

The vi.mock declaration uses './github-utils' but the dynamic imports below use './github-utils.js'. These resolve to the same module in the current build setup, but make them consistent to avoid subtle mock-bypass issues under different module resolution modes.

Recommended fix: Change both dynamic imports to import('./github-utils') (no .js), matching the mock declaration.

File: apps/worker/src/lib/github-PR-tracker.test.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  vi.mocked(getInstallationOctokit).mockResolvedValue({
+    paginate: mockPaginate,
+    request: mockRequest,
+  } as never)
+
+  return { mockPaginate, mockRequest }
+}
+
+describe('ingestRepoMergedPRs', () => {
+  beforeEach(() => {
```
Comment (kelvinchen8) [inline apps/worker/src/lib/github-PR-tracker.test.ts]: Redundant beforeEach duplicates what the global setup already does
What's wrong: The global setup file (vitest.setup.ts) already registers beforeEach(() => vi.clearAllMocks()). The test file adds another identical beforeEach at line 62, so clearAllMocks runs twice per test with no effect.

vitest.setup.ts already calls vi.clearAllMocks() in a global beforeEach. This local one is redundant and can be removed.

Recommended fix: Remove the beforeEach block from the test file.

File: apps/worker/src/lib/github-PR-tracker.test.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  it('passes the correct date range to the GitHub search query', async () => {
+    const { mockPaginate } = setupOctokit([[]], [])
+
+    await ingestRepoMergedPRs(repo, weekStart, weekEnd)
+
+    const query = mockPaginate.mock.calls[0][1].q
+    expect(query).toContain('merged:2026-04-28T00:00:00.000+00:00..2026-05-04T23:59:59.000+00:00')
+  })
+
+  it('upserts the same PR twice without error', async () => {
```
Comment (kelvinchen8) [inline apps/worker/src/lib/github-PR-tracker.test.ts]: Idempotency test makes no meaningful assertion
What's wrong: The "upserts the same PR twice without error" test only checks that the second call resolves to 1, which is already covered by the earlier "upserts a merged PR" test. It doesn't verify that db.pRFact.upsert was called a second time, or that the upsert where clause correctly identifies the existing record. As written it adds no coverage.

This test only asserts .resolves.toBe(1), which is already verified by the previous test. To actually test idempotency, assert that db.pRFact.upsert was called twice total (once per invocation) with the same where key both times, confirming the upsert path is hit rather than insert-with-conflict.

Recommended fix: After both calls, assert:


expect(db.pRFact.upsert).toHaveBeenCalledTimes(2)
expect(db.pRFact.upsert).toHaveBeenNthCalledWith(2, expect.objectContaining({
  where: { repoId_number: { repoId: 'repo-1', number: 1 } },
}))

**Review by kelvinchen8 — CHANGES_REQUESTED** [review]

File: apps/worker/src/lib/github-PR-tracker.integration.test.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
+import { describe, it, expect, vi, beforeEach } from 'vitest'
+import { db } from '@repo/db'
+import { getInstallationOctokit } from '../lib/github-auth'
```
Comment (kelvinchen8) [inline apps/worker/src/lib/github-PR-tracker.integration.test.ts]: `import { getInstallationOctokit } from './github-auth'`

no need for ../lib

File: apps/worker/src/lib/github-PR-tracker.integration.test.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    const count = await ingestRepoMergedPRs(repo, weekStart, weekEnd)
+
+    expect(count).toBe(0)
+    expect(await db.pRFact.count()).toBe(0)
+  })
+
+  it('writes the PR to the DB with correct fields', async () => {
+    const repo = await seedRepo()
+    setupOctokit([[{ number: 1 }], []], [makeFullPr()])
```
Comment (kelvinchen8) [inline apps/worker/src/lib/github-PR-tracker.integration.test.ts]: do we need the extra empty list? we only need one entry right?

same for line 100, 113, 132, 147, 150

**Review by kelvinchen8 — CHANGES_REQUESTED** [review]

File: apps/worker/src/lib/github-PR-tracker.integration.test.ts (hunk 5 above)
Comment (kelvinchen8) [reply apps/worker/src/lib/github-PR-tracker.integration.test.ts]: hm is this related to this:

github-commit-tracker.ts imports getInstallationOctokit via ../lib/github-auth, but this test mocks ./github-auth. In Vitest, mocks are matched by module ID, so this can fail to intercept the dependency and cause the real getInstallationOctokit to run (which will throw due to missing GITHUB_APP_ID/GITHUB_APP_PRIVATE_KEY). Mock the same specifier used by the module under test (e.g. vi.mock('../lib/github-auth')) and import from that same path when configuring vi.mocked(...).mockResolvedValue(...).

a copilot comment in [joe's pr](https://github.com/UoaWDCC/projects-health-dashboard/pull/143/changes)

**Review by kelvinchen8 — APPROVED** [review]
lgtm
